src/callouts.rs

```diff
@@ -0,0 +1,354 @@
+            if !st.is_empty() {
+                let s = format!("{}: {}", &sname, st);
+
+                if !self.use_syslog {
+                    eprint!("{}", &s);
+                } else {
+                    let _ = logging::journal_print(Priority::Warning, &s);
+                }
+            }
+        }
+        if stdout {
+            let st = String::from_utf8_lossy(&output.stdout);
+            if !st.is_empty() {
+                let s = format!("{}: {}", &sname, st);
+
+                if !self.use_syslog {
+                    print!("{}", &s);
+                } else {
+                    let _ = logging::journal_print(Priority::Notice, &s);
+                }
+            }
+        }
+    }
+
+    fn invoke_first_matching_script<P: AsRef<Path>>(
+        &self,
+        dev: &mut MDev,
+        dir: P,
+        event: Event,
+        action: Action,
+    ) -> Option<(PathBuf, Output)> {
+        if dir.as_ref().read_dir().ok()?.count() == 0 {
+            return None;
+        }
+
+        for s in dir.as_ref().read_dir().ok()? {
+            let path = s.ok()?.path();
+
+            match self.invoke_script(dev, &path, event, action).ok() {
+                Some(res) => {
+                    if res.status.code().is_none() {
+                        warn!("Script was terminated by a signal");
+                        continue;
+                    } else if res.status.code() != Some(2) {
+                        return Some((path, res));
+                    } else {
+                        debug!(
+                            "Device type {} unmatched by callout script",
+                            dev.mdev_type().ok()?
+                        );
+                    }
+                }
+                _ => {
+                    debug!("failed to execute callout script {:?}", path);
+                    continue;
+                }
+            }
+        }
+        None
+    }
+
+    fn callout(&mut self, dev: &mut MDev, event: Event, action: Action) -> Result<()> {
+        let dir = dev.env.callout_script_base();
+
+        if !dir.is_dir() {
+            return Ok(());
+        }
+
+        let rc;
+
+        if let Some(ref s) = self.script {
+            let output = self.invoke_script(dev, s, event, action)?;
+            self.print_output(&output, s, true, false);
+            rc = output.status.code();
+        } else {
+            match self.invoke_first_matching_script(dev, dir, event, action) {
+                Some((path, output)) => {
+                    self.print_output(&output, &path, true, false);
+                    self.script = Some(path);
+                    rc = output.status.code();
+                }
+                None => rc = None,
+            }
+        }
+
+        match rc {
+            Some(0) | None => Ok(()),
+            _ => Err(anyhow!(
+                "aborting command due to results from callout script {:?}",
+                self.script.as_ref().unwrap()
```

Comment:
Inside of this function, you don't actually know whether the command will be aborted or not. The caller (`Callout::invoke()`) is the one who decides whether to abort the command or not. So perhaps we should use a slightly different message. For example: 

```suggestion
            Some(n) => Err(anyhow!(
                "callout script {:?} failed with return code {}",
                self.script.as_ref().unwrap(), n
```

(adding the return code to the error message might help script writers debug failures, but it might not be the best user experience...)